### ISSUE TITLE
ldap2pg seems to ignore ldaprc files

### ISSUE THREAD

**Issue by sgns12** [dalibo/ldap2pg#272]
Hi
I describe here 3 cases that dry run runs fine , but 
1. no ldaprc anywhere and runs 
2. bad ldaprc file and it runs
in both above case empty BIND DN is passwd, but all works
3. define unix env variables and all works fine and BIND DN is not empty
question is how ldap2pg even know to which ldap server to connect, let alone runs fine?
thank you
shimshon
run on centos 7.5 with openldap

1st case:
No env variables set and no ldaprc file

``` console
postgres@itsrhv22112:/tmp/112> set |grep LDAP
postgres@itsrhv22112:/tmp/112> ls -ltr
total 20
-rw------- 1 postgres postgres 4816 Apr 10 16:42 test.yml.112
-rw------- 1 postgres postgres  142 Apr 18 10:58 ldaprc-ignored
-rw------- 1 postgres postgres  164 Apr 18 15:10 export.profile.112
-rw------- 1 postgres postgres   65 Apr 18 15:13 unset.profile.112
postgres@itsrhv22112:/tmp/112> ldap2pg -c test.yml.112  -C  --dry -v
[ldap2pg.config        INFO] Starting ldap2pg 4.18.
[ldap2pg.config       DEBUG] Trying test.yml.112.
[ldap2pg.config        INFO] Using /tmp/112/test.yml.112.
[ldap2pg.config       WARNI] Unknown config entry: postgres:dbaownrole_query.
[ldap2pg.config       DEBUG] Read check from argv.
[ldap2pg.config       DEBUG] Read dry from argv.
[ldap2pg.config       DEBUG] Read verbosity from argv.
[ldap2pg.config       DEBUG] Read postgres:databases_query from YAML.
[ldap2pg.config       DEBUG] Read postgres:managed_roles_query from YAML.
[ldap2pg.config       DEBUG] Read postgres:schemas_query from YAML.
[ldap2pg.config       DEBUG] Read privileges from YAML.
[ldap2pg.config       DEBUG] Read sync_map from YAML.
[ldap2pg.config       DEBUG] Configuration loaded.
[ldap2pg.script       DEBUG] Connecting to LDAP directory.
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/etc/ldap/ldap.conf'
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/usr/local/postgres/ldaprc'
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/usr/local/postgres/.ldaprc'
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/tmp/112/ldaprc'
[ldap2pg.ldap         DEBUG] Read BINDDN from YAML.
[ldap2pg.ldap         DEBUG] Read REFERRALS from YAML.
[ldap2pg.ldap         DEBUG] Read URI from YAML.
[ldap2pg.ldap         DEBUG] Read HOST from YAML.
[ldap2pg.ldap         DEBUG] Read PASSWORD from YAML.
[ldap2pg.ldap         DEBUG] Read PORT from YAML.
[ldap2pg.ldap         DEBUG] Connecting to LDAP server ldap://:389.
[ldap2pg.ldap         DEBUG] Trying simple bind.
[ldap2pg.ldap         DEBUG] Doing: ldapwhoami -x
...............
[ldap2pg.manager      DEBUG] Found GRANT ddl on testdb.public for * to dbaownrole (full) from YAML.
[ldap2pg.manager       INFO] Querying LDAP ou=11-TEST,ou=testdb,ou=... (cn=itsrhv22...
[ldap2pg.ldap         DEBUG] Doing: ldapsearch -x -b ou=11-TEST,ou=testdb,ou=readrole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com -s sub '(cn=itsrhv22112)' member
[ldap2pg.manager      DEBUG] Got 1 entries from LDAP.
[ldap2pg.manager      DEBUG] Found role r555555r from cn=itsrhv22112,ou=11-TEST,ou=testdb,ou=readrole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com.
```

all works fine w/o ldaprc and ldap log shows
Apr 18 16:10:36 itsrhv22112 slapd[29574]: conn=1041 op=0 BIND dn="" method=128

2nd case :
now I do have ldaprc file - but it is wrong ldaprc since I point to bogus ldap and bad password
but it works.

``` console
postgres@itsrhv22112:/tmp/112> cat ldaprc
uri ldap://bogus.it.statestr.com:2389
binddn "uid=PostgresUser,ou=system,ou=users,dc=statestreet,dc=com"
password Passowrd

and all runs fine too  - it notices now that I have ldaprc but seems to ignore it and it is bad 
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/etc/ldap/ldap.conf'
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/usr/local/postgres/ldaprc'
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/usr/local/postgres/.ldaprc'
[ldap2pg.ldap         DEBUG] Found rcfile /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read uri from /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read binddn from /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read password from /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read BINDDN from YAML.
[ldap2pg.ldap         DEBUG] Read REFERRALS from YAML.
[ldap2pg.ldap         DEBUG] Read URI from YAML.
[ldap2pg.ldap         DEBUG] Read HOST from YAML.
[ldap2pg.ldap         DEBUG] Read PASSWORD from YAML.
[ldap2pg.ldap         DEBUG] Read PORT from YAML.
[ldap2pg.ldap         DEBUG] Connecting to LDAP server ldap://:389.
..........
[ldap2pg.manager       INFO] Querying LDAP ou=11-TEST,ou=testdb,ou=... (cn=itsrhv22...
[ldap2pg.ldap         DEBUG] Doing: ldapsearch -x -b ou=11-TEST,ou=testdb,ou=readrole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com -s sub '(cn=itsrhv22112)' member
[ldap2pg.manager      DEBUG] Got 1 entries from LDAP.
[ldap2pg.manager      DEBUG] Found role r555555r from cn=itsrhv22112,ou=11-TEST,ou=testdb,ou=readrole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com.
```

all works fine and again I get in ldap.log empty BIND DN
srhv22112 slapd[29574]: conn=1042 op=0 BIND dn="" method=128

3rd case - which at least I understand why it works.
now I set env variables

``` console
postgres@itsrhv22112:/tmp/112> . export.profile.112
postgres@itsrhv22112:/tmp/112> set|grep LDAP
LDAPBINDDN=ou='IFSUsers,dc=it,dc=statestr,dc=com'
LDAPPASSWORD=D0ntDo1t
LDAPPORT=389
LDAPURI=ldap://itsrhv22112.it.statestr.com:389
and again it all runs fine but picks up env varialbles

[ldap2pg.script       DEBUG] Connecting to LDAP directory.
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/etc/ldap/ldap.conf'
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/usr/local/postgres/ldaprc'
[ldap2pg.ldap         DEBUG] Ignoring: [Errno 2] No such file or directory: u'/usr/local/postgres/.ldaprc'
[ldap2pg.ldap         DEBUG] Found rcfile /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read uri from /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read binddn from /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read password from /tmp/112/ldaprc.
[ldap2pg.ldap         DEBUG] Read BINDDN from env.
[ldap2pg.ldap         DEBUG] Read PASSWORD from env.
[ldap2pg.ldap         DEBUG] Read PORT from env.
[ldap2pg.ldap         DEBUG] Read URI from env.
[ldap2pg.ldap         DEBUG] Read BINDDN from YAML.
[ldap2pg.ldap         DEBUG] Read REFERRALS from YAML.
[ldap2pg.ldap         DEBUG] Read URI from YAML.
[ldap2pg.ldap         DEBUG] Read HOST from YAML.
[ldap2pg.ldap         DEBUG] Read PASSWORD from YAML.
[ldap2pg.ldap         DEBUG] Read PORT from YAML.
[ldap2pg.ldap         DEBUG] Connecting to LDAP server ldap://itsrhv22112.it.statestr.com:389.
[ldap2pg.ldap         DEBUG] Trying simple bind.
[ldap2pg.ldap         DEBUG] Doing: ldapwhoami -x -D ou=IFSUsers,dc=it,dc=statestr,dc=com -W
................
[ldap2pg.manager       INFO] Querying LDAP ou=11-TEST,ou=testdb,ou=... (cn=itsrhv22...
[ldap2pg.ldap         DEBUG] Doing: ldapsearch -x -D ou=IFSUsers,dc=it,dc=statestr,dc=com -W -b ou=11-TEST,ou=testdb,ou=readrole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com -s sub '(cn=itsrhv22112)' member
[ldap2pg.manager      DEBUG] Got 1 entries from LDAP.

take a look that now ldapwhoami passws full BIND DN  -D u=IFSUsers,dc=it,dc=statestr,dc=com 
and we see it in ldap.log
Apr 18 16:22:14 itsrhv22112 slapd[29574]: conn=1044 op=0 BIND dn="ou=IFSUsers,dc=it,dc=statestr,dc=com" method=128
Apr 18 16:22:14 itsrhv22112 slapd[29574]: conn=1044 op=0 BIND dn="ou=IFSUsers,dc=it,dc=statestr,dc=com" mech=SIMPLE ssf=0
```

**Comment by sgns12** [comment]
sorry, forgot yml
postgres@itsrhv22112:/tmp/112> cat test.yml.112
#
# **LDAP2PG SAMPLE CONFIGURATION**
#
# This is a sample starting point configuration file for ldap2pg.yml. Including
# static roles, groups, privilege and LDAP query.
#
# This configuration assumes the following principles:
#
# - All LDAP users are grouped in `ldap_roles` group.
# - Read privileges are granted to `readrole` group.
# - Write privileges are granted to `writerole` group.
# - DDL privileges are granted to `dbaownrole` group.
# - We have one or more databases with public and maybe a schema.
# - Grants are not specific to a schema. Once you're writer in a database, you
#   are writer to all schemas in it.
#
# Adapt to your needs! See also full documentation on how to configure ldap2pg
# at https://ldap2pg.readthedocs.io/en/latest/config/.
#

verbosity: 5

postgres:

  # Disable roles introspection by setting query to null
# roles_query: null

  # Scope the database where to purge objects when dropping roles. This is the
  # scope of grant on `__all__` databases.
  databases_query: [ testdb ]
  # List of managed schema. This skip pg_toast, pg_temp1, etc. but not pg_catalog.
  schemas_query: |
    SELECT nspname FROM pg_catalog.pg_namespace
    WHERE nspname = 'pg_catalog' OR nspname NOT LIKE 'pg_%'
  # Return managed roles which can be dropped or revoked.
  managed_roles_query: |
    SELECT 'public'
    UNION
    SELECT DISTINCT role.rolname
    FROM pg_roles AS role
    LEFT OUTER JOIN pg_auth_members AS ms ON ms.member = role.oid
    LEFT OUTER JOIN pg_roles AS ldap_roles
      ON ldap_roles.rolname = 'ldap_roles' AND ldap_roles.oid = ms.roleid
    WHERE role.rolname IN ('ldap_roles', 'readrole', 'writerole', 'dbaownrole')
        OR ldap_roles.oid IS NOT NULL
    ORDER BY 1;

  # Since readrole/writer/dbaownrole groups are globals, we have a global
  # dbaownrole_query.
  dbaownrole_query: |
    SELECT DISTINCT role.rolname
    FROM pg_catalog.pg_roles AS role
    JOIN pg_catalog.pg_auth_members AS ms ON ms.member = role.oid
    JOIN pg_catalog.pg_roles AS dbaownrole
      ON dbaownrole.rolname = 'dbaownrole' AND dbaownrole.oid = ms.roleid
    ORDER BY 1;



privileges:
  # Define an privilege group `ro` with read-only grants
  ro:
  - __connect__
  - __execute__
  - __select_on_tables__
  - __select_on_sequences__
  - __usage_on_schemas__
  - __usage_on_types__

  # `rw` privilege group lists write-only grants
  rw:
  - __temporary__
  - __all_on_tables__
  - __all_on_sequences__

  # `ddl` privilege group lists DDL only grants.
  ddl:
  - __create_on_schemas__


sync_map:
# First, setup static roles and grants
- roles:
  - names:
    - ldap_roles
    - readrole
    options: NOLOGIN
    comment: Custom static comment.
  - name:
    - writerole
    # Grant reading to writerole
    parent: readrole
    options: NOLOGIN
  - name: dbaownrole
    # Grant read/write to dbaownrole
    parent: writerole
    options: NOLOGIN
  # Now grant privileges to each groups
  grant:
  - privilege: ro
    role: readrole
    # Let's everyone see pg_catalog
    schema: __all__
  - privilege: rw
    role: writerole
    # But avoid writerole to write in pg_catalog
    schema: public
  # Allow ddl to create tables in public only
  - privilege: ddl
    role: dbaownrole
    schema: public
  # dbaownrole must have write access to pg_catalog
  - privilege: rw
    role: dbaownrole
    schema: pg_catalog
  # Grants on specific schema testdb.public:
  - privilege: rw
    role: writerole
    database: testdb
    schema: public
  - privilege: ddl
    role: dbaownrole
    database: testdb
    schema: public

# Now query LDAP to create roles and grant them privileges by parenting.

#--------------------------------------------------------------------
- ldap:
  # base: ou=groups,dc=ldap,dc=ldap2pg,dc=docker
    base: ou=11-TEST,ou=testdb,ou=readrole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com
    filter: "(cn=itsrhv22112)"
  role:
    name: '{member.cn}'
    options: LOGIN
    parent:
    - ldap_roles
    - readrole
    comment: "Custom comment from LDAP: {dn}"
#--------------------------------------------------------------------
- ldap:
  # base: ou=groups,dc=ldap,dc=ldap2pg,dc=docker
    base: ou=11-TEST,ou=testdb,ou=writerole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com
    filter: "(cn=itsrhv22112)"
  role:
    name: '{member.cn}'
    options: LOGIN
    parent:
    - ldap_roles
    - writerole
    comment: "Custom comment from LDAP: {dn}"
#--------------------------------------------------------------------
- ldap:
  # base: ou=groups,dc=ldap,dc=ldap2pg,dc=docker
    base: ou=11-TEST,ou=testdb,ou=dbarole,ou=PG_Group_OU,ou=Group,dc=it,dc=statestr,dc=com
    filter: "(cn=itsrhv22112)"
  role:
    name: '{member.cn}'
    options: LOGIN SUPERUSER
    parent:
    - ldap_roles
    - dbaownrole
    comment: "Custom comment from LDAP: {dn}"
postgres@itsrhv22112:/tmp/112>

**Comment by sgns12** [comment]
P.S.
also ldap.conf points to different ldap server for all clients, that I don't use.
have 'unknown' to ldap.con ldap server.
so have no idea how it finds 'my test' ldap server.

**Comment by sgns12** [comment]
i figured out why it works despite not having ldaprc file - 
ldap2pg goes by default to localhost port 389, even w/o binddn and works.

now, i do have another case where ldap server is on another machine and i do have ldaprc file
but does not work w/o setting unix variables
postgres@itsrhv22112:/tmp/dev> cat ldaprc
uri: ldap://agsos137.it.statestr.com:2389
binddn: "uid=PostgresUser,ou=system,ou=users,dc=statestreet,dc=com"
#user e468014
password: HappyNow12

i did strace and see ldap2pg see ldaprc file
getcwd("/tmp/dev", 1026)                = 9
open("/etc/ldap/ldap.conf", O_RDONLY)   = -1 ENOENT (No such file or directory)
write(2, "\33[0;37m[ldap2pg.ldap         DEB"..., 110) = 110
open("/usr/local/postgres/ldaprc", O_RDONLY) = -1 ENOENT (No such file or directory)
write(2, "\33[0;37m[ldap2pg.ldap         DEB"..., 117) = 117
open("/usr/local/postgres/.ldaprc", O_RDONLY) = -1 ENOENT (No such file or directory)
write(2, "\33[0;37m[ldap2pg.ldap         DEB"..., 118) = 118
open("/tmp/dev/ldaprc", O_RDONLY)       = 3
fstat(3, {st_dev=makedev(0, 37), st_ino=504139061, st_mode=S_IFREG|0600, st_nlink=1, st_uid=90, st_gid=90, st_blksize=4096, st_blocks=8, st_size=145, st_atime=2019/04/22-12:16:25.067738799, st_mtime=2019/04/22-12:16:14.848490722, st_ctime=2019/04/22-12:16:22.020664829}) = 0
write(2, "\33[0;37m[ldap2pg.ldap         DEB"..., 70) = 70
read(3, "uri: ldap://agsos137.it.statestr"..., 8192) = 145

but ldap2pg does not call that ldap, instead it calls  local ldap
connect(3, {sa_family=AF_INET, sin_port=htons(389), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
 
 but when i set these variables in unix all is ok and it goes to the correct ldap server
export LDAPBINDDN="uid=PostgresUser,ou=system,ou=users,dc=statestreet,dc=com"
export LDAPPORT=2389
export LDAPPASSWORD=HappyNow12
export LDAPURI=ldap://agsos137.it.statestr.com:2389


connect(3, {sa_family=AF_INET, sin_port=htons(2389), sin_addr=inet_addr("10.1.71.7")}, 16) = 0
PING agsos137.it.statestr.com (10.1.71.7) 56(84) bytes of data.
64 bytes from agsos137.statestr.com (10.1.71.7): icmp_seq=1 ttl=251 time=0.689 ms

thank you

**Comment by bersace** [comment]
Thanks for the detailed report, i'll investigate this.

**Comment by bersace** [comment]
Hi @sgns12 , verbose logs looks to tell the precedence of origin. Last value wins. The thing that may be strange is that there is **two** namespaces for configuration LDAP: native libldap and ldap2pg. ldap2pg.yml has precedence over libldap-like settings. Here is the full order, last wins:

- defaults
- /etc/ldap/ldap.conf
- ~/ldaprc
- ~/.ldaprc
- ./ldaprc
- file at LDAPCONF
- file at LDAPRC
- ldap2pg.yml
- LDAP* envvars.

The logs are a bit odd, for the following reason. LDAP* env vars are loaded twice. Once to override ldap2pg.yml and a second time to read some more LDAP* vars like LDAPRC and LDAPCONF. Thus, when you define LDAPBINDDN, you have this in the logs:

```
[ldap2pg.ldap         DEBUG] Read BINDDN from env.
[ldap2pg.ldap         DEBUG] Read BINDDN from YAML.
```

Actually, the YAML is fed by the same envvar. Actually, the final behaviour should be straigthforward or need a fix. Only the --verbose logs are a bit odd.

Does this explanation helps you understand what's going on ?

**Comment by bersace** [comment]
Now, can you explain me a way to reproduce the bug with:

1. steps to reproduce
2. expected result
3. bad result returned

Thanks !

**Comment by bersace** [comment]
@sgns12 I got it. The bug is fixed in #275. Thanks for pointing this !!

**Comment by bersace** [comment]
@sgns12 Would you mind closing this ticket once you're ok ?
